File: rust/541.reverse-string-ii.rs
/*
 * @lc app=leetcode id=541 lang=rust
 *
 * [541] Reverse String II
 */

// @lc code=start
impl Solution {
    pub fn reverse_str(s: String, k: i32) -> String {
        s.chars()
            .collect::<Vec<char>>()
            .chunks_mut(2 * k as usize)
            .map(|cs| {
                if cs.len() > k as usize {
                    cs[0..k as usize].reverse();
                } else {
                    cs.reverse()
                }
                cs.iter().collect::<String>()
            })
            .collect::<String>()
    }
}
// @lc code=end

File: javascript/541.reverse-string-ii.js
/*
 * @lc app=leetcode id=541 lang=javascript
 *
 * [541] Reverse String II
 */

// @lc code=start
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var reverseStr = function (s, k) {
  let str = "";
  let chars = s.split('');
  while (chars.length > 0) {
    const slice = chars.splice(0, 2 * k);
    str +=
      slice.length > k
        ? slice.splice(0, k).reverse().join("") + slice.join("")
        : slice.reverse().join("");
  }
  return str;
};
// @lc code=end
